// eslint-plugin-jest-formatting/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const padding_1 = require("./rules/padding");
const paddingConfigs = {
    afterAll: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 1,
        },
        {
            paddingType: 1,
            prevStatementType: 1,
            nextStatementType: 0,
        },
    ],
    afterEach: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 2,
        },
        {
            paddingType: 1,
            prevStatementType: 2,
            nextStatementType: 0,
        },
    ],
    beforeAll: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 3,
        },
        {
            paddingType: 1,
            prevStatementType: 3,
            nextStatementType: 0,
        },
    ],
    beforeEach: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 4,
        },
        {
            paddingType: 1,
            prevStatementType: 4,
            nextStatementType: 0,
        },
    ],
    describe: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: [
                5,
                7,
                11,
            ],
        },
        {
            paddingType: 1,
            prevStatementType: [
                5,
                7,
                11,
            ],
            nextStatementType: 0,
        },
    ],
    expect: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 6,
        },
        {
            paddingType: 1,
            prevStatementType: 6,
            nextStatementType: 0,
        },
        {
            paddingType: 0,
            prevStatementType: 6,
            nextStatementType: 6,
        },
    ],
    test: [
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: [
                10,
                9,
                8,
                12,
                13,
            ],
        },
        {
            paddingType: 1,
            prevStatementType: [
                10,
                9,
                8,
                12,
                13,
            ],
            nextStatementType: 0,
        },
    ],
};
exports.rules = {
    'padding-around-after-all-blocks': padding_1.createRule(paddingConfigs.afterAll),
    'padding-around-after-each-blocks': padding_1.createRule(paddingConfigs.afterEach),
    'padding-around-before-all-blocks': padding_1.createRule(paddingConfigs.beforeAll),
    'padding-around-before-each-blocks': padding_1.createRule(paddingConfigs.beforeEach),
    'padding-around-describe-blocks': padding_1.createRule(paddingConfigs.describe),
    'padding-around-expect-groups': padding_1.createRule(paddingConfigs.expect),
    'padding-around-test-blocks': padding_1.createRule(paddingConfigs.test),
    'padding-around-all': padding_1.createRule([].concat(...Object.keys(paddingConfigs).map(k => paddingConfigs[k]))),
    'padding-before-after-all-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 1,
        },
    ], true),
    'padding-before-after-each-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 2,
        },
    ], true),
    'padding-before-before-all-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 3,
        },
    ], true),
    'padding-before-before-each-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 4,
        },
    ], true),
    'padding-before-describe-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 5,
        },
    ], true),
    'padding-before-expect-statements': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: 6,
        },
        {
            paddingType: 0,
            prevStatementType: 6,
            nextStatementType: 6,
        },
    ], true),
    'padding-before-test-blocks': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: [10, 9],
        },
    ], true),
    'padding-before-all': padding_1.createRule([
        {
            paddingType: 1,
            prevStatementType: 0,
            nextStatementType: [
                1,
                2,
                3,
                4,
                5,
                6,
                9,
                10,
            ],
        },
        {
            paddingType: 0,
            prevStatementType: 6,
            nextStatementType: 6,
        },
    ], true),
};
exports.configs = {
    recommended: {
        plugins: ['jest-formatting'],
        overrides: [
            {
                files: [
                    '**/*.test.*',
                    '**/*_test.*',
                    '**/*Test.*',
                    '**/*.spec.*',
                    '**/*_spec.*',
                    '**/*Spec.*',
                    '**/__tests__/*',
                ],
                rules: {
                    'jest-formatting/padding-around-after-all-blocks': 2,
                    'jest-formatting/padding-around-after-each-blocks': 2,
                    'jest-formatting/padding-around-before-all-blocks': 2,
                    'jest-formatting/padding-around-before-each-blocks': 2,
                    'jest-formatting/padding-around-describe-blocks': 2,
                    'jest-formatting/padding-around-test-blocks': 2,
                },
            },
        ],
    },
    strict: {
        plugins: ['jest-formatting'],
        overrides: [
            {
                files: [
                    '**/*.test.*',
                    '**/*_test.*',
                    '**/*Test.*',
                    '**/*.spec.*',
                    '**/*_spec.*',
                    '**/*Spec.*',
                    '**/__tests__/*',
                ],
                rules: {
                    'jest-formatting/padding-around-all': 2,
                },
            },
        ],
    },
};
